File: src/collab/validate_draft.py
"""
Validate a draft markdown file has required fields and correct format.

Usage:
  corrkit validate-draft drafts/2026-02-19-example.md
  corrkit validate-draft drafts/*.md
"""

import argparse
import re
import sys
from pathlib import Path

META_RE = re.compile(r"^\*\*(.+?)\*\*:\s*(.+)$", re.MULTILINE)

REQUIRED_FIELDS = {"To"}
RECOMMENDED_FIELDS = {"Status", "Author"}
VALID_STATUSES = {"draft", "review", "approved", "sent"}


def validate_draft(path: Path) -> list[str]:
    """Validate a draft file. Returns list of issues (empty = valid)."""
    issues: list[str] = []

    if not path.exists():
        return [f"File not found: {path}"]

    text = path.read_text(encoding="utf-8")
    lines = text.split("\n")

    # Check for subject heading
    has_subject = any(line.startswith("# ") for line in lines)
    if not has_subject:
        issues.append("Missing subject: no '# Subject' heading found")

    # Parse metadata fields
    meta: dict[str, str] = {}
    for m in META_RE.finditer(text):
        meta[m.group(1)] = m.group(2).strip()

    # Required fields
    for field in REQUIRED_FIELDS:
        if field not in meta:
            issues.append(f"Missing required field: **{field}**")

    # Recommended fields (warn, don't error)
    for field in RECOMMENDED_FIELDS:
        if field not in meta:
            issues.append(f"Warning: missing recommended field: **{field}**")

    # Status validation
    status = meta.get("Status", "").lower()
    if status and status not in VALID_STATUSES:
        issues.append(
            f"Invalid status '{meta['Status']}'. "
            f"Valid: {', '.join(sorted(VALID_STATUSES))}"
        )

    # Check status is 'review' (common mistake: leaving as 'draft')
    if status == "draft":
        issues.append(
            "Warning: Status is 'draft'. Set to 'review' when ready for Brian"
        )

    # Check for --- separator
    has_separator = any(line.strip() == "---" for line in lines)
    if not has_separator:
        issues.append("Missing '---' separator between metadata and body")

    # Check body exists after separator
    if has_separator:
        sep_idx = next(i for i, line in enumerate(lines) if line.strip() == "---")
        body = "\n".join(lines[sep_idx + 1 :]).strip()
        if not body:
            issues.append("Warning: empty body after --- separator")

    return issues


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Validate draft markdown files"
    )
    parser.add_argument(
        "files",
        nargs="+",
        help="Draft markdown file(s) to validate",
    )
    args = parser.parse_args()

    paths = [Path(p) for p in args.files]
    all_ok = True

    for path in paths:
        issues = validate_draft(path)
        if issues:
            all_ok = False
            errors = [i for i in issues if not i.startswith("Warning:")]
            warnings = [i for i in issues if i.startswith("Warning:")]
            print(f"{path}:")
            for issue in errors:
                print(f"  ERROR: {issue}")
            for issue in warnings:
                print(f"  {issue}")
            print()
        else:
            print(f"{path}: OK")

    sys.exit(0 if all_ok else 1)


if __name__ == "__main__":
    main()

File: src/mailbox/validate_draft.rs
//! Validate draft markdown files.

use anyhow::Result;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashMap;
use std::path::{Path, PathBuf};

static META_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?m)^\*\*(.+?)\*\*:\s*(.+)$").unwrap());

const REQUIRED_FIELDS: &[&str] = &["To"];
const RECOMMENDED_FIELDS: &[&str] = &["Status", "Author"];
const VALID_STATUSES: &[&str] = &["draft", "review", "approved", "sent"];

/// Validate a draft file. Returns list of issues (empty = valid).
pub fn validate_draft(path: &Path) -> Vec<String> {
    let mut issues = Vec::new();

    if !path.exists() {
        return vec![format!("File not found: {}", path.display())];
    }

    let text = match std::fs::read_to_string(path) {
        Ok(t) => t,
        Err(e) => return vec![format!("Cannot read {}: {}", path.display(), e)],
    };
    let lines: Vec<&str> = text.split('\n').collect();

    // Check for subject heading
    let has_subject = lines.iter().any(|line| line.starts_with("# "));
    if !has_subject {
        issues.push("Missing subject: no '# Subject' heading found".to_string());
    }

    // Parse metadata fields
    let mut meta: HashMap<String, String> = HashMap::new();
    for cap in META_RE.captures_iter(&text) {
        meta.insert(cap[1].to_string(), cap[2].trim().to_string());
    }

    // Required fields
    for field in REQUIRED_FIELDS {
        if !meta.contains_key(*field) {
            issues.push(format!("Missing required field: **{}**", field));
        }
    }

    // Recommended fields (warn, don't error)
    for field in RECOMMENDED_FIELDS {
        if !meta.contains_key(*field) {
            issues.push(format!(
                "Warning: missing recommended field: **{}**",
                field
            ));
        }
    }

    // Status validation
    let status = meta
        .get("Status")
        .map(|s| s.to_lowercase())
        .unwrap_or_default();
    if !status.is_empty() && !VALID_STATUSES.contains(&status.as_str()) {
        issues.push(format!(
            "Invalid status '{}'. Valid: {}",
            meta.get("Status").unwrap(),
            VALID_STATUSES.to_vec().join(", ")
        ));
    }

    if status == "draft" {
        issues.push(
            "Warning: Status is 'draft'. Set to 'review' when ready for review".to_string(),
        );
    }

    // Check for --- separator
    let has_separator = lines.iter().any(|line| line.trim() == "---");
    if !has_separator {
        issues.push("Missing '---' separator between metadata and body".to_string());
    }

    // Check body exists after separator
    if has_separator {
        if let Some(sep_idx) = lines.iter().position(|line| line.trim() == "---") {
            let body: String = lines[sep_idx + 1..].join("\n");
            if body.trim().is_empty() {
                issues.push("Warning: empty body after --- separator".to_string());
            }
        }
    }

    issues
}

/// corky validate-draft FILE [FILE...]
pub fn run(files: &[PathBuf]) -> Result<()> {
    let mut all_ok = true;

    for path in files {
        let issues = validate_draft(path);
        if !issues.is_empty() {
            all_ok = false;
            let errors: Vec<_> = issues
                .iter()
                .filter(|i| !i.starts_with("Warning:"))
                .collect();
            let warnings: Vec<_> = issues.iter().filter(|i| i.starts_with("Warning:")).collect();
            println!("{}:", path.display());
            for issue in errors {
                println!("  ERROR: {}", issue);
            }
            for issue in warnings {
                println!("  {}", issue);
            }
            println!();
        } else {
            println!("{}: OK", path.display());
        }
    }

    if !all_ok {
        std::process::exit(1);
    }
    Ok(())
}
